### PR TITLE
Reset accounts

### LINKED ISSUE #29: Add Account Recovery
User should be able to choose to recover an account from mnemonics on initial account creation, instead of creating a new account

### PR DESCRIPTION
**Pull request by tommyjk21**
Adds a menu item allowing the user to reset their account. closes #29

---

Before we can merge this PR, please make sure that all the following items have been checked off:

- [ ] Linked to an issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Added **one** line describing your change in [`CHANGELOG.md`](https://github.com/manta-network/manta-signer/blob/main/CHANGELOG.md) and added the appropriate `changelog` label to the PR.
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -36,18 +36,18 @@ derivative = { version = "2.2.0", default-features = false, features = ["use_cor
...
-manta-pay = { git = "https://github.com/manta-network/manta-rs", tag = "v0.5.1", default-features = false, features = ["bs58", "groth16", "serde", "wallet"] }
-manta-util = { git = "https://github.com/manta-network/manta-rs", tag = "v0.5.1", default-features = false }
+manta-accounting = { git = "https://github.com/manta-network/manta-rs", branch = "fix/reduce-key-rederivations", default-features = false, features = ["cocoon-fs"] }
+manta-crypto = { git = "https://github.com/manta-network/manta-rs", branch = "fix/reduce-key-rederivations", default-features = false, features = ["getrandom"] }
+manta-parameters = { git = "https://github.com/manta-network/manta-rs", branch = "fix/reduce-key-rederivations", default-features = false, features = ["download"] }
+manta-pay = { git = "https://github.com/manta-network/manta-rs", branch = "fix/reduce-key-rederivations", default-features = false, features = ["bs58", "groth16", "serde", "wallet"] }
+manta-util = { git = "https://github.com/manta-network/manta-rs", branch = "fix/reduce-key-rederivations", default-features = false }
```
Comment (Kevingislason) [inline Cargo.toml:43]: I think you can use `tag = "v0.5.1"` instead of specifying branch now

File: ui/src-tauri/src/main.rs [hunk 2]
```diff
@@ -221,38 +237,50 @@ fn main() {
...
         Config::try_default().expect("Unable to generate the default server configuration.");
 
+    let (reset_tx, reset_rx) = channel(1);
+
     let mut app = tauri::Builder::default()
         .system_tray(
             SystemTray::new().with_menu(
                 SystemTrayMenu::new()
                     .add_item(CustomMenuItem::new("about", "About"))
+                    .add_item(CustomMenuItem::new("reset", "Reset Seed"))
```
Comment (Kevingislason) [inline ui/src-tauri/src/main.rs:247]: "Reset Account" is a little more obvious imo

File: ui/src/pages/ResetAccount.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+          <Header className="recovery-phrase-header">
+            Enter Recovery Phrase
+          </Header>
+          <div className="recovery-phrase-warning">
+            ⚠️  Never share your recovery phrase with anyone! ⚠️ 
+          </div>
+          {!validSeedPhrase && <Label basic color='yellow'>Enter a valid seed phrase</Label> }
+          {validSeedPhrase && <Label basic color='blue'>Cool! Hit send!</Label> }
+          <Form>
+              <Form.TextArea onChange={onEnterSeedPhrase}>
```
Comment (Kevingislason) [inline ui/src/pages/ResetAccount.js:58]: for some reason I can't copy-paste into this text area

File: ui/src/pages/ResetAccount.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    <>
+      {recoveryPrompt && (
+        <>
+          <Header className="recovery-phrase-header">
+            Enter Recovery Phrase
+          </Header>
+          <div className="recovery-phrase-warning">
+            ⚠️  Never share your recovery phrase with anyone! ⚠️ 
+          </div>
+          {!validSeedPhrase && <Label basic color='yellow'>Enter a valid seed phrase</Label> }
```
Comment (Kevingislason) [inline ui/src/pages/ResetAccount.js:55]: I don't think this should display if the user hasn't typed in anything yet

**Review by Kevingislason — CHANGES_REQUESTED** [review]
When the user clicks "Reset account", or when the user opens the app without a seed, I think they should get the option to either recover from mnemonic or restart from a fresh seed. So I think we need one more page with two buttons--"New account" and "Recover account", which direct the user to the appropriate page. This page could also include a warning--as it stands, users might accidentally delete their accounts on a misclick.

edit based on discussion in full stack meeting:
we can do "Reset account" in the upper menu bar--we don't need to do this now, can do this in a separate PR. We still want to include a warning before allowing the user to reset their account.

File: ui/src-tauri/src/main.rs (hunk 2 above)
Comment (Kevingislason) [reply ui/src-tauri/src/main.rs:247]: Actually "Recover from seed" is best

File: src/service.rs [hunk 5]
```diff
@@ -82,6 +85,9 @@ pub enum Error {
     /// Failed to Load SDK Parameters
     ParameterLoadingError,
 
+    /// Channel closed Error
+    ChannelClosedError,
```
Comment (bhgomes) [inline src/service.rs:89]:
```suggestion
    /// Reset Channel Closed Error
    ResetChannelClosedError,
```

File: src/secret.rs [hunk 6]
```diff
@@ -27,6 +27,9 @@ pub use password_hash::{Error as PasswordHashError, PasswordHasher, PasswordVeri
 pub use secrecy::{ExposeSecret, Secret, SecretString};
 pub use subtle::{Choice, ConstantTimeEq, CtOption};
 
+/// mnemonic, password
+pub type ResetInfo = (SecretString, SecretString);
```
Comment (bhgomes) [inline src/secret.rs:31]: Make this a struct

File: examples/test_server.rs [hunk 7]
```diff
@@ -58,5 +59,6 @@ async fn main() -> Result<(), Error> {
     if let Some(url) = std::env::args().nth(1) {
         config.service_url = url;
     }
-    service::start(config, MockUser::new(&mut OsRng)).await
+    let (_tx, rx) = channel(1);
+    service::start(config, MockUser::new(&mut OsRng), rx).await
```
Comment (bhgomes) [inline examples/test_server.rs:63]:
```suggestion
    service::start(config, MockUser::new(&mut OsRng), None).await
```

File: src/service.rs [hunk 8]
```diff
@@ -393,9 +403,77 @@ where
...
+            info!("Reset channel closed, exiting.").expect("Logger error in reset handler");
+            break;
+        }
+    }
+}
+
 /// Starts the signer server with `config` and `authorizer`.
 #[inline]
-pub async fn start<A>(config: Config, authorizer: A) -> Result<()>
+pub async fn start<A>(config: Config, authorizer: A, reset_rx: Receiver<ResetInfo>) -> Result<()>
```
Comment (bhgomes) [inline src/service.rs:476]:
```suggestion
pub async fn start<A>(config: Config, authorizer: A, reset_receiver: Option<Receiver<ResetInfo>>) -> Result<()>
```

Comment (bhgomes) [reply src/service.rs:476]: And add documentation about this optional handler.

File: src/service.rs [hunk 9]
```diff
@@ -408,7 +486,21 @@ where
...
+    task::spawn(async move {
+        handle_reset(reset_rx, init_rx, conf).await;
+    });
+
+    let server = Server::build(config, authorizer).await?;
+
+    if init_tx.send(server.state_handle()).await.is_err() {
+        return Err(Error::ChannelClosedError);
+    }
+
```
Comment (bhgomes) [inline src/service.rs:502]: Move this into its own function and use full names for all variables.

File: src/service.rs [hunk 10]
```diff
@@ -393,9 +403,77 @@ where
...
+                let _ = do_reset(mnemonic, password, &config).await;
+            }
+        }
+    };
+
+    loop {
+        if let Some((mnemonic, password)) = reset_rx.recv().await {
+            let mut new_state = do_reset(mnemonic, password, &config).await;
+
+            std::mem::swap(&mut *state_handle.lock(), &mut new_state);
```
Comment (bhgomes) [inline src/service.rs:466]: Don't use `std` when the same function exists in `core`. Always prefer this ordering: `core` -> `alloc` -> `std`.

File: src/service.rs [hunk 11]
```diff
@@ -232,6 +238,10 @@ where
         })
     }
 
+    fn state_handle(&self) -> Arc<Mutex<State>> {
+        self.state.clone()
+    }
```
Comment (bhgomes) [inline src/service.rs:243]:
```suggestion
    /// Returns a handle to the state of the server.
    #[inline]
    pub fn state(&self) -> Arc<Mutex<State>> {
        self.state.clone()
    }
```

File: src/service.rs [hunk 12]
```diff
@@ -393,9 +403,77 @@ where
     Ok(Body::from_json(&f().await?)?.into())
 }
 
+async fn do_reset(mnemonic: SecretString, password: SecretString, config: &Config) -> State {
```
Comment (bhgomes) [inline src/service.rs:406]: Always prefer non-transient arguments first: 

```suggestion
async fn reset_state(config: &Config, mnemonic: SecretString, password: SecretString) -> State {
```

File: src/service.rs [hunk 13]
```diff
@@ -393,9 +403,77 @@ where
...
+    task::spawn_blocking(move || File::save(&path, &password_hash_bytes, cloned_state))
+        .await
+        .expect("Signer state save failed")
+        .expect("Could not save signer state.");
+    let signer = Signer::from_parts(parameters, state);
+
+    State {
+        config: config.clone(),
+        signer,
+    }
```
Comment (bhgomes) [inline src/service.rs:442]:
```suggestion
    State {
        config: config.clone(),
        signer: Signer::from_parts(parameters, state),
    }
```

File: src/service.rs (hunk 12 above)
Comment (bhgomes) [inline src/service.rs:406]: Please use proper error handling for this function. Don't just expect on all the errors, pass them up to the caller.

File: src/service.rs [hunk 14]
```diff
@@ -393,9 +403,77 @@ where
...
+    );
+    info!("saving signer state").expect("Logger error in reset handler");
+    let password_hash_bytes = password_hash.as_bytes();
+    let cloned_state = state.clone();
+    let path = config.data_path.clone();
+
+    task::spawn_blocking(move || File::save(&path, &password_hash_bytes, cloned_state))
+        .await
+        .expect("Signer state save failed")
+        .expect("Could not save signer state.");
```
Comment (bhgomes) [inline src/service.rs:436]: Much of this code is repeated in other parts of `service.rs`. Please find a refactoring that does not introduce repetition (even if that means writing new functions that generalize over all the use cases).

File: ui/src-tauri/src/main.rs [hunk 15]
```diff
@@ -200,6 +203,19 @@ impl PasswordStore {
...
+#[tauri::command]
+async fn send_recovery_info(
+    reset_handle: State<'_, ResetHandle>,
+    phrase: String,
+    password: String,
+) -> Result<bool, ()> {
+    let info = (Secret::new(phrase), Secret::new(password));
+
+    reset_handle.send(info).await.expect("Reset channel broke");
+    Ok(true)
```
Comment (bhgomes) [inline ui/src-tauri/src/main.rs:216]: Add a `from_strings` (bikeshed the name) method to the `ResetInfo` struct that takes `Strings` and converts them into `Secret`s. Also, return the error back up to the caller.

```suggestion
    reset_handle.send(ResetInfo::from_strings(phrase, password)).await
```

File: ui/src-tauri/src/main.rs [hunk 16]
```diff
@@ -221,38 +237,50 @@ fn main() {
     let config =
         Config::try_default().expect("Unable to generate the default server configuration.");
 
+    let (reset_tx, reset_rx) = channel(1);
```
Comment (bhgomes) [inline ui/src-tauri/src/main.rs:240]:
```suggestion
    let (reset_sender, reset_receiver) = channel(1);
```

File: ui/src-tauri/src/main.rs [hunk 17]
```diff
@@ -221,38 +237,50 @@ fn main() {
...
         )
         .on_system_tray_event(move |app, event| {
             if let SystemTrayEvent::MenuItemClick { id, .. } = event {
                 match id.as_str() {
                     "about" => app.get_window("about").unwrap().show().unwrap(),
+                    "reset" => {
+                        let config = app.state::<Config>().inner().clone();
+                        fs::remove_file(config.data_path).expect("Failed to remove file");
+                        let window = app.get_window("main").unwrap();
+                        window.emit("reset", "").unwrap();
```
Comment (bhgomes) [inline ui/src-tauri/src/main.rs:259]:
```suggestion
                        app.get_window("main").unwrap().emit("reset", "").unwrap();
```

File: ui/src-tauri/src/main.rs [hunk 18]
```diff
@@ -221,38 +237,50 @@ fn main() {
...
         .manage(config)
         .setup(|app| {
             let window = app.get_window("main").unwrap();
             let config = app.state::<Config>().inner().clone();
             let password_store = app.state::<PasswordStore>().handle();
+
             spawn(async move {
                 let (password, retry) = password_store.into_channel().await;
-                service::start(config, User::new(window, password, retry))
+                service::start(config, User::new(window, password, retry), reset_rx)
```
Comment (bhgomes) [inline ui/src-tauri/src/main.rs:276]:
```suggestion
                service::start(config, User::new(window, password, retry), Some(reset_receiver))
```

File: ui/src-tauri/src/main.rs [hunk 19]
```diff
@@ -221,38 +237,50 @@ fn main() {
...
+                        let window = app.get_window("main").unwrap();
+                        window.emit("reset", "").unwrap();
+                    }
                     "exit" => app.exit(0),
                     _ => {}
                 }
             }
         })
         .manage(PasswordStore::default())
+        .manage(reset_tx)
```
Comment (bhgomes) [inline ui/src-tauri/src/main.rs:267]:
```suggestion
        .manage(reset_sender)
```

File: ui/src/App.js [hunk 20]
```diff
@@ -58,6 +64,20 @@ function App() {
...
+    console.log("[INFO]: Setup listener.");
+    listen('reset', (event) => {
+      console.log("[INFO]: Wake: ", event);
+      setCurrentPage(RESET_PAGE);
+      appWindow.show();
+    });
+  };
+
+  const sendRecoveryInfo = async (phrase, password) => {
+    console.log("[INFO]: Send info to signer server.");
```
Comment (bhgomes) [inline ui/src/App.js:77]:
```suggestion
    console.log("[INFO]: Send reset info to signer server.");
```

**Review by bhgomes — CHANGES_REQUESTED** [review]

**Pull request closed without merging**